README.adoc
= Test App
Doc Writer <jpr@ethinking.de>
Developed under NodeJs 7.7.1

== Installation

First clone or download this project on your computer.

To work with this app, you need to have `nodeJs` intalled on your machine. If you don't, go to https://nodejs.org/en/[nodeJs] website and install the last recommended version.

Then you need to install `gulp` globally on your machine. Open a terminal and enter: `npm i gulp -g`.

=== Dependencies installation:

Go inside the `frontend-test` folder, open your terminal here and enter `npm i`. This will isntall all the dependencies required to use this app.

Once finished, enter `gulp serve` to start the local server. It should open a page in your browser. If not, open the page `http://localhost:3000/index`.

=== Exercises:

1- In the `Design` folder, you will find 4 jpg files. I will ask you to create the menu like it is on the design. It's a responsive menu with just one breakpoint. It should look similar as it is on the design, but no need to be "pixel perfect".

Write your HTML code in the file located in `app/views/header/header.ejs`. Don't put any html or body tags inside, they are already somewhere else in the template, write only your code for the menu. You can write normal HTML in this file, don't be scared of the .ejs extension, it changes nothing.

Write your CSS in the header.scss file located in `app/sass/header/header.scss`.

Write your JAVASCRIPT in the menu.js file located in `app/js/main/menu.js`.

The icons are located inside `public/images`, to display them, just call: `./images/YOUR_IMAGE_NAME`. Nav icons are different than on the design, it's not a problem.

Any change on your code will refresh the page. The CSS will be refreshed without refreshing the all page.

jQuery is present in the project so you can use it. No other library is allowed (no Bootstrap for example).
Usage of css flex will be a plus, but not mandatory.

2- For the second exercise, open the `app/js/main/ex2.js` file, and answer the few questions.

=== Send your results:

Once all done, put the results into a personnal GIT repository, and send me the link to it (add jpr@ethinking.de). If you don't know how to do this, just zip and send me by mail what's inside the `app` folder.

I wish you good luck, anf if you have any question or problem, please contact Jérôme @ jpr@ethinking.de (In English preferably)



CHANGELOG.adoc
= Change Log
Doc Writer <jpr@ethinking.de>

All notable changes to this project will be documented in this file.

The format is based on http://keepachangelog.com/[Keep a Changelog]
and this project adheres to http://semver.org/[Semantic Versioning].

== [3.1.0] - 2018-01-08

=== Modified
. Changed livereload by browsersync

'''

== [3.0.2] - 2017-11-06

=== Fixed
. Removed "mainFolder" into `cdm.js`.
. We don't clean anymore the `dist` folder.
. Removed "workingSite into `gulp.js`.
. Other small fixes

=== Future improvements
. Generate css and js in .tmp folder and then move it to public or build

'''

== [3.0.1] - 2017-11-01

=== Modified
. Several small bugs fixes

'''

== [3.0.0] - 2017-10-24

=== Modified
. All generated .js and .css files in the public folder are now ignored by GIT, the build is not (in the dist folder at the root of the project)
. Back to single app

'''

== [2.0.0] - 2017-07-18

=== New
. Clickdummy app goes multi-site!
. Multi-site with a common folder.
. Possibility to add group templates.
. Possibility to rename sass files.
. Possibility to create all sass files from a BEM name.
. Possibility to work with scss and not only sass now.
. More ES6!!!

'''

== [1.0.5] - 2017-05-12

=== Modified
. Vendor javascript are now in a separate file in `js/vendor/vendor.js`
. Custom css are now in a `main` folder
. Gulp.js modified to listen to the new `vendor.js` file and update the one in the `dist` folder (_it was only possible to update it when updating a custom js file before_)
. Removed `babel` from package.json + gulp
. Put in `package.jon dependencies` all we need to just generate the assets we need in the dist folder, and in `devdependencies` what we need for development.
. Improved the documentation
. Removed the hello message when a new page is created
. renamed `index.js` file for command line functionalities by `cmd.js`

'''

== [1.0.4] - 2017-05-10

=== Added
. Added actual url to the data sent to the pages (useful for navigation to display the active item)

=== Fixed
. Removed stringification followed by par parsing of the json.

=== Modified
. Page generated contains the code to display views instead of the hello message.
. Created a `gulp build` task
. Added `vendor-scripts` to gulp watch for js

'''

== [1.0.3] - 2017-04-28

=== Fixed
. Gulp update: css and js are now correctly rendered in `app/css` and `app/js`
. Added `concat` to gulp for js files (only the last file was uglyfied)

'''

== [1.0.2] - 2017-04-24

=== Changed
. All files for the app are now in the `app` folder

'''

== [1.0.1] - 2017-04-18

=== Added
. CHANGELOG.adoc

=== Changed
. README.adoc

=== Fixed
. Updated "_partials_" name into .json files and cli by "_views_"

'''

== [1.0.0] - 2017-04-04
=== First commit


app.js
var express = require( 'express' );
var path = require( 'path' );
var favicon = require( 'serve-favicon' );
var logger = require( 'morgan' );
var cookieParser = require( 'cookie-parser' );
var bodyParser = require( 'body-parser' );
var routes = require( './routes/index' );

var app = express();

app.use( '/', routes );

var env = process.env.NODE_ENV || 'development';
app.locals.ENV = env;
app.locals.ENV_DEVELOPMENT = env == 'development';

// view engine setup

app.set( 'views', path.join( __dirname, 'routes' ) );
app.set( 'view engine', 'ejs' );

// app.use(favicon(__dirname + '/dist/img/favicon.ico'));
app.use( logger( 'dev' ) );
app.use( bodyParser.json() );
app.use( bodyParser.urlencoded( {
	extended: true
} ) );
app.use( cookieParser() );
app.use( express.static( path.join( __dirname, 'app/public' ) ) );

/// catch 404 and forward to error handler
app.use( function( req, res, next ) {
	var err = new Error( 'Not Found' );
	err.status = 404;
	next( err );
} );

/// error handlers

// development error handler
// will print stacktrace

if( app.get( 'env' ) === 'development' ) {
	app.use( function( err, req, res, next ) {
		res.status( err.status || 500 );
		res.render( 'error', {
			message: err.message,
			error  : err,
			title  : 'error'
		} );
	} );
}

// production error handler
// no stacktraces leaked to user
app.use( function( err, req, res, next ) {
	res.status( err.status || 500 );
	res.render( 'error', {
		message: err.message,
		error  : {},
		title  : 'error'
	} );
} );

module.exports = app;


routes/index.js
const express = require( 'express' );
const app = express();
const request = require( 'request' );
const router = express.Router();
const fs = require( 'fs' );
var path = require( 'path' );

app.use( express.static( 'public' ) );

router.get( '/:page', function( req, res ) {

	var viewData = {
		pageName   : req.url,
		headerData : fs.readFileSync( 'data/header.json', 'utf-8', function( err, data ) {
			return JSON.parse( data );
		} ),
		contentData: fs.readFileSync( 'data/' + req.params.page + '.json', 'utf-8', function( err, data ) {
			return JSON.parse( data );
		} ),
		footerData : fs.readFileSync( 'data/footer.json', 'utf-8', function( err, data ) {
			return JSON.parse( data );
		} )
	};

	res.render( '../app/views/index', viewData, function( err, html ) {

		if( err ) {
			if( err.message.indexOf( 'Failed to lookup view' ) !== -1 ) {
				res.render( "error", {
					message: err.message,
					error  : { status: err.status, stack: err.stack }
				} );
			}
			throw err;
		}
		res.send( html );

	} );
} );

module.exports = router;

app/js/main/ex2.js
// Data array
let dataArray = [
	{ firstname: 'Jan', name: 'Müller', age: 35 },
	{ firstname: 'David', name: 'Johnson', age: 24 },
	{ firstname: 'John', name: 'Papa', age: 43 },
	{ firstname: 'Philip', name: 'Georges', age: 19 },
	{ firstname: 'Christian', name: 'Laval', age: 30 }
];

// For all answers, just log the result: console.log()

// Write a script to check if someone in the dataArray
// is more than 40 years old (script should return true or false)
// It should work with any array with the same structure



// Make an array containing only the names inside
// Result should be: [ 'Müller', 'Johnson', 'Papa', 'Georges', 'Laval']



// Make from data array a new array with only the persons less than 35 years old inside
// The new array should look like this at the end :
/*
[
 { firstname: 'David', name: 'Johnson', age: 24 },
 { firstname: 'Philip', name: 'Georges', age: 19 },
 { firstname: 'Christian', name: 'Laval', age: 30 }
 ];
 */




gulpfile.js
const gulp = require( 'gulp' );
const nodemon = require( 'gulp-nodemon' );
const gulpLoadPlugins = require( 'gulp-load-plugins' );
const del = require( 'del' );
const browserSync = require( 'browser-sync' ).create();
const gulpsync = require( 'gulp-sync' )( gulp );          // For sync tasks
const $ = gulpLoadPlugins();
const message = require( 'gulp-message' );                // used for debugging
const babel = require( 'gulp-babel' );

gulp.task( 'sass', function() {
//	message.debug( 'sass' );
	gulp.src( './app/sass/*.*' )
	    .pipe( $.plumber() )
	    .pipe( $.sourcemaps.init() )
	    .pipe( $.sass.sync().on( 'error', $.sass.logError ) )
	    .pipe( $.autoprefixer( { browsers: [ 'last 2 versions' ] } ) )
	    .pipe( gulp.dest( '.tmp/css' ) )
	    .pipe( gulp.dest( 'app/public/css' ) )
	    .pipe( $.sass( { outputStyle: 'compressed' } ) )
	    .pipe( $.rename( { suffix: '.min' } ) )
	    .pipe( $.sourcemaps.write( './' ) )
	    .pipe( gulp.dest( '.tmp/css' ) )
	    .pipe( gulp.dest( './app/public/css' ) )
	    .pipe( browserSync.stream() );
} );

gulp.task( 'clean', del.bind( null, [ '.tmp', '.dist' ] ) );

gulp.task( 'watch', function() {
	gulp.watch( 'app/sass/**/*.*', [ 'sass' ] );
	gulp.watch( 'app/js/main/*.js', [ 'scripts', 'scripts-min' ] );
	gulp.watch( 'app/js/vendor/vendor.js', [ 'vendor-scripts' ] );
} );

gulp.task( 'scripts', function() {
	return gulp.src( 'app/js/main/*.js' )
	           .pipe( babel( {
		                         presets: [ 'env' ]
	                         } ) )
	           .pipe( $.plumber() )
	           .pipe( $.sourcemaps.init() )
	           .pipe( $.concat( 'main.js' ) )
	           .pipe( $.sourcemaps.write( '.' ) )
	           .pipe( gulp.dest( '.tmp/js' ) )
	           .pipe( gulp.dest( './app/public/js' ) );
} );

gulp.task( 'scripts-min', function() {
	return gulp.src( 'app/js/main/*.js' )
	           .pipe( babel( {
		                         presets: [ 'env' ]
	                         } ) )
	           .pipe( $.plumber() )
	           .pipe( $.sourcemaps.init() )
	           .pipe( $.concat( 'main.min.js' ) )
	           .pipe( $.uglify() )
	           .pipe( $.sourcemaps.write( '.' ) )
	           .pipe( gulp.dest( '.tmp/js' ) )
	           .pipe( gulp.dest( './app/public/js' ) );
} );

gulp.task( 'vendor-scripts', function() {
	const vendorfile = require( './app/js/vendor/vendor.js' );
	const vendorJs = vendorfile.vendorJs;
	gulp.src( vendorJs )
	    .pipe( $.plumber() )
	    .pipe( $.concat( 'vendor.js' ) )
	    .pipe( gulp.dest( '.tmp/js' ) )
	    .pipe( $.rename( 'vendor.min.js' ) )
	    .pipe( $.uglify() )
	    .pipe( gulp.dest( '.tmp/js' ) )
	    .pipe( gulp.dest( './app/public/js' ) );
} );

gulp.task( 'nodemon', function( cb ) {
	let called = false;
	return nodemon(
		{
			script: 'bin/www',
			ext   : 'js ejs coffee',
			stdout: false
		} )
	.on( 'start', function onStart() {
		// ensure start only got called once
		if( !called ) {
			called = true;
			cb();
		}
	} )
	.on( 'restart', function onRestart() {
		// reload connected browsers after a slight delay
		setTimeout( function reload() {
			browserSync.reload(
				{
					stream: false
				} );
		}, 500 );
	} );
} );

gulp.task( 'browser-sync', [ 'nodemon' ], function() {

	// for more browser-sync config options: http://www.browsersync.io/docs/options/
	browserSync.init(
		{
			proxy        : 'http://localhost:4000/index',
			// informs browser-sync to use the following port for the proxied app
			// notice that the default port is 4000, which would clash with our expressjs
			port         : 3000,
			injectChanges: true,
			reloadDelay  : 250
		} );
} );

gulp.task( 'copy-to-dist', function() {
	gulp.src( [ 'app/public/**/*', '.tmp/**/*' ] )
	    .pipe( gulp.dest( 'dist' ) );
} );

gulp.task( 'copy-to-public', function() {
	gulp.src( [ '.tmp/**/*' ] )
	    .pipe( gulp.dest( './app/public' ) );
} );

gulp.task( 'reload', function() {
	browserSync.reload();
} );

gulp.task( 'serve', gulpsync.sync(
	[
		'clean',
		'sass',
		'scripts',
		'scripts-min',
		'vendor-scripts',
		'copy-to-public',
		'browser-sync',
		'watch'
	] ) );

gulp.task( 'build', gulpsync.sync(
	[
		'clean',
		'sass',
		'scripts',
		'scripts-min',
		'vendor-scripts',
		'copy-to-dist'
	] ) );

gulp.task( 'default', [
	'serve'
] );
